### PR TITLE
implement host lookup support

### PR DESCRIPTION
**Pull request by flokli**
This adds support for the GETAI, GETHOSTBYADDR, GETHOSTBYADDRv6, GETHOSTBYNAME, GETHOSTBYNAMEv6 request types.

For the more complex GETAI lookup, we use the dns_lookup crate.

In previous iterations of this change, we also used the same underlying getaddrinfo call to respond to the  gethostbyname/gethostbyaddr operations.

Even though gethostbyname/gethostbyaddr officially are deprecated, there's a lot of tools still using it, and relying on them behaving differently.

So it's important to still implement it, with exactly the same behaviour, to prevent some tools from breaking (like `hostname --fqdn`, see https://github.com/nix-community/nsncd/issues/4).

We FFI the right Glibc gethostbyname_2r/gethostbyaddr_2r (now deprecated) functions and use it to back the GETHOSTBYNAME, GETHOSTBYNAME6, GETHOSTBYADDR, and GETHOSTBYADDR6 Nscd interfaces.

Took us three tries to get this right. This is actually the third full rewrite.

The Nscd behaviour for these two legacy functions is *really* confusing. We're supposed to ignore the herrno (herrno != errno!!) and set it to 0 if gethostbyaddr/name returns a non-null hostent. If we end up with a null hostent, we return the herrno together with a dummy hostent header.

I tried to keep things as safe as possible by extracting the glibc hostent to a proper rust structure. This structure mirrors the libc hostent in a Rust idiomatic way. We should probably try to upstream the FFI part of this commit to the Nix crate at some point.

Fixes https://github.com/nix-community/nsncd/issues/4

### CODE REVIEW

**Comment by flokli** [comment]
We tested these changes extensively using [sockburp](https://github.com/flokli/sockburp/wiki/glibc%E2%80%90nscd---nsncd-on-NixOS) on the [host-lookups](https://github.com/nix-community/nsncd/tree/host-lookups) branch on the `nix-community` fork. This is a cherry-pick of the latest iteration.

We'll give it some more testing on our side, but wanted to send this out already, so you can take a look.

**Comment by flokli** [comment]
Marking as ready for review, PTAL :-)

**Comment by picnoir** [comment]
There's something still off with the getaddrinfo canonical name.

**Comment by flokli** [comment]
The remaining issue is fixed too, and cherry-picked here - see https://github.com/nix-community/nsncd/pull/10 for details.

File: src/handlers.rs
Comment (leifwalsh) [inline src/handlers.rs:153]: Why `SOCK_STREAM` here? I guess, what if the nscd client is trying to do `SOCK_DGRAM`, what do we do and what would real `nscd` do?

File: src/handlers.rs
Comment (leifwalsh) [inline src/handlers.rs:508]:
```suggestion
/// 4. addrslen: int32. Total length of the IPv4/6 adresses we're
```

File: src/handlers.rs
Comment (leifwalsh) [inline src/handlers.rs:510]:
```suggestion
/// 5. canonlen: int32. Total length of the null-terminated canonical
```

File: src/handlers.rs
Comment (leifwalsh) [inline src/handlers.rs:520]: I think you need a 9, the canonical name, nul-terminated.

File: src/handlers.rs
Comment (leifwalsh) [inline src/handlers.rs:208]:
```suggestion
            let hostent = gethostbyaddr_r(LibcIp::V4(address_bytes))?;
```

I probably would have written this, and just let the original error surface and get logged. But I don't have strong (or generally good) opinions about error handling. I'll defer to @blinsay if he wants to have a stronger or better opinion.

(same idea below too)

File: src/handlers.rs
Comment (leifwalsh) [inline src/handlers.rs:420]: Do you need the type twice?

File: src/handlers.rs
Comment (leifwalsh) [inline src/handlers.rs:440]: I think this could be simplified by adding a `HST_RESPONSE_HEADER_NOT_FOUND` like you did for `AI_RESPONSE_HEADER_NOT_FOUND` - get rid of all the `has_addrs` checks. Unless `aliases` can still be valid or something?

File: src/handlers.rs
Comment (leifwalsh) [inline src/handlers.rs:463]: Where does `4 * 8` come from? Is that really just the size of `HstResponseHeader`?

File: src/handlers.rs
Comment (leifwalsh) [inline src/handlers.rs:480]:
```suggestion
            buf.extend_from_slice(buf_aliases_size.as_slice());
```

You can do this unconditionally, right? `buf_aliases_size` should be the empty vector if it's empty? (Same about `buf_aliases` below)

**Review by leifwalsh — CHANGES_REQUESTED** [review]
A couple minor typos and some questions. Overall I think it's fairly straightforward (or as straightforward as code dealing with getaddrinfo and friends can be). Haven't tried running it at all but the code feels like it would be debuggable if something crashed?

File: src/ffi.rs
Comment (picnoir) [reply src/ffi.rs:97]: Absolutely. Makes sense.

File: src/ffi.rs
Comment (picnoir) [reply src/ffi.rs:202]: Empirical trial-and-error. I tried several gethostbyxxx requests and saw that most of them fit in a 200 bytes buffer. 

Looking [at](https://sourcegraph.com/github.com/NetBSD/src/-/blob/lib/libc/net/gethnamaddr.c?L1332) [sourcegraph](https://sourcegraph.com/github.com/NetBSD/src/-/blob/distrib/utils/libhack/gethost.c?L79) , it seems like there's no consensus wrt. the buffer size and we're in the lower range with 200 bytes (also, lots of ppl seem to disregard the ERANGE errno…).

Every time we resize the buffer, we have to go through another gethostbyxxx call, which is quite expensive. Better start with a size that fits most requests.

File: src/ffi.rs
Comment (picnoir) [reply src/ffi.rs:218]: Good point. I guess that past 500M, we should be more than done. Let's use that arbitrary limit.

File: src/ffi.rs
Comment (picnoir) [reply src/ffi.rs:226]: > What does "The stream" refer to?

I have honestly no idea what I mean here :O This + the extra line makes me think this comment has been butchered during a git conflict resolution at some point. Let's rewrite it.

File: src/ffi.rs
Comment (picnoir) [reply src/ffi.rs:237]: It's nonsensical and outdated :)

File: src/ffi.rs
Comment (picnoir) [reply src/ffi.rs:248]: We shouldn't, thanks.

(I rewrote this binding 3 times, we used to error out when the return value was !=0. Which is not what the nscd client expects…)

File: src/handlers.rs
Comment (picnoir) [reply src/handlers.rs:152]: Clarified the comment:

```rust
                // Setting socktype to SOCK_STREAM to deduplicate the
                // results. If we don't do that, glibc gives us twice
                // every entry: once as a SOCK_STREAM, another one as
                // SOCK_DGRAM.
```

File: src/ffi.rs
Comment (flokli) [reply src/ffi.rs:218]: I'd probably use 1..10M as an upper limit already. That's quite a lot of hostnames / IPs ;-)

File: src/ffi.rs
Comment (picnoir) [reply src/ffi.rs:202]: On the Nscd side, the scratch buffer is initialized at 1024 bytes:

```
struct scratch_buffer {
  void *data;    /* Pointer to the beginning of the scratch area.  */
  size_t length; /* Allocated space at the data pointer, in bytes.  */
  union { max_align_t __align; char __c[1024]; } __space;
};
```

Let's re-use this default instead of this 200.

File: src/ffi.rs
Comment (picnoir) [reply src/ffi.rs:247]: :+1:  Outdated comment.

File: src/handlers.rs
Comment (picnoir) [reply src/handlers.rs:463]: Yeah, that's the header. Rewrote that more explicitly using `std::mem::size_of`

File: src/handlers.rs
Comment (picnoir) [reply src/handlers.rs:153]: Added

```rust
                // There's no way to convey socktype in the the Nscd
                // protocol, neither in the request nor response.
                //
                // Set this to SOCK_STREAM to match glibc and unscd
                // behaviour.
```

Does it make sense?

**Comment by picnoir** [comment]
I squashed the nits/small modifications in the original commits.

Separated the getaddrinfo serialization routine (asked in https://github.com/twosigma/nsncd/pull/71#discussion_r1375506047) in a separate commit: it's a lot, it'll likely be simpler to review it from a separate commit.

File: src/handlers.rs
Comment (picnoir) [reply src/handlers.rs:212]: Did not address this one yet. Let me know what you prefer.

File: src/handlers.rs
Comment (picnoir) [reply src/handlers.rs:532]: Added.

File: src/handlers.rs
Comment (flokli) [reply src/handlers.rs:157]: @leifwalsh please mark as resolved if you're ok with this.

**Comment by picnoir** [comment]
(fixed the clippy error)

File: src/ffi.rs
Comment (leifwalsh) [inline src/ffi.rs:228]:
```suggestion
/// af is either nix::libc::AF_INET or nix::libc::AF_INET6
```

**Review by leifwalsh — CHANGES_REQUESTED** [review]

**Comment by leifwalsh** [comment]
@picnoir have you sent us a [CLA](https://github.com/twosigma/nsncd/blob/main/cla/TSOS%20CLA%20-%20Individual%20-%20nsncd.pdf) yet? I'm trying to track this down internally too but people are on vacation/doing end-of-year activities so I figured I'd just ask you while I wait to hear back. If you haven't, please do so (emailing a scanned PDF is fine) and let me know.

**Comment by leifwalsh** [comment]
Otherwise LGTM, @geofft @blinsay let me know if you want to take a pass, otherwise I'll merge once we have CLAs or on Wednesday, whichever comes first.

**Comment by flokli** [comment]
> @flokli and @picnoir have you sent us a [CLA](https://github.com/twosigma/nsncd/blob/main/cla/TSOS%20CLA%20-%20Individual%20-%20nsncd.pdf) yet? I'm trying to track this down internally too but people are on vacation/doing end-of-year activities so I figured I'd just ask you while I wait to hear back. If you haven't, please do so (emailing a scanned PDF is fine) and let me know.
> 
> (I think we accepted one of @flokli's earlier PRs as de minimis, so we probably don't have one from either of you yet)

I'm very sure I sent one out Nov 14 2022 (to tsos at twosigma.com` just found the email), and am somewhat convinced @picnoir did so as well at around the same time.

**Comment by picnoir** [comment]
Yup, according to https://github.com/twosigma/nsncd/discussions/44#discussioncomment-4135685 I did.

File: src/ffi.rs
Comment (flokli) [inline src/ffi.rs:86]: serialize_hostent

File: src/handlers.rs
Comment (flokli) [inline src/handlers.rs:391]: To match the rest of the structure of this file, I'd probably change this to be a
```rust
fn serialize_hostent(hostent: Hostent) -> Result<Vec<u8>> {
  // …
}
```

File: src/handlers.rs
Comment (flokli) [inline src/handlers.rs:533]: This could also just take the `AiResponse` by value (consume it), so there's less cloning needed.

**Review by leifwalsh — APPROVED** [review]

File: src/ffi.rs
Comment (blinsay) [inline src/ffi.rs:174]: want to move this to its own function? maybe something like `Hostent::error_value` would be nice

File: src/handlers.rs
Comment (blinsay) [reply src/handlers.rs:212]: The whole thing with the sentinel error-case hostent getting returned from `gethostbyaddr_r` is very odd to me. We don't have to reproduce libc's strangeness exactly, even if we do have to serialize that error value and send it back to the caller.

What if we did something like:

```rust
enum HostentError {
  NotFound,
  TryAgain,
  NoRecovery,
  NoData,
  Other(anyhow::Error),
}

pub fn gethostbyname2_r(name: String, af: libc::c_int) -> Result<Hostent, HostentError> {
  ...
}
```

and then in handlers we could do something like this:

```rust
            let hostent = match gethostbyaddr_r(LibcIp::V4(address_bytes)) {
                Ok(hostent) => hostent,
                Err(Other(e)) => bail!("oh no, something very bad happened: {e}"),
                Err(e) => {
                  log::debug!("gethostbyaddr_r(..) = {e:?}");
                  Hostent::error_value()
                },
            };
```

File: src/ffi.rs
Comment (blinsay) [inline src/ffi.rs:244]: I think by the book this may need to be `MaybeUninit<u8>` even though it's never read. Let me consult with @geofft.

File: src/handlers.rs
Comment (blinsay) [inline src/handlers.rs:428]: it's a bit confusing that this check happens after the aliases have been serialized.

File: src/handlers.rs
Comment (blinsay) [inline src/handlers.rs:491]: can this case get moved way up to the beginning of the function? it's a bit lost dangling down here:

```rust
if hostent.addr_list.is_empty() {
  return Ok(protocol::HstReponseHeader{ .... })
}

// rest of the fn goes here
```

File: src/handlers.rs
Comment (blinsay) [inline src/handlers.rs:539]: nit: this can be `b_addrs.extend(ip.octets())`

File: src/handlers.rs
Comment (blinsay) [inline src/handlers.rs:546]: any reason this doesn't use `octets`?

https://doc.rust-lang.org/stable/std/net/struct.Ipv6Addr.html#method.octets

**Review by blinsay — COMMENTED** [review]
Thanks for taking three(!) passes at this, there's a ton here.

File: src/ffi.rs
Comment (blinsay) [reply src/ffi.rs:244]: We chatted and I think I'm being overly cautious here.

File: src/ffi.rs
Comment (flokli) [reply src/ffi.rs:111]: Added a check.

File: src/ffi.rs
Comment (flokli) [reply src/ffi.rs:124]: Ack.

File: src/ffi.rs
Comment (flokli) [reply src/ffi.rs:113]: Ack.

File: src/ffi.rs
Comment (flokli) [reply src/ffi.rs:132]: Updated the comments.

File: src/ffi.rs
Comment (flokli) [reply src/ffi.rs:174]: Done, thanks for the suggestion.

File: src/handlers.rs
Comment (flokli) [reply src/handlers.rs:428]: Moved before that.

File: src/handlers.rs
Comment (flokli) [reply src/handlers.rs:491]: Moved that return further up, allowed simplifying some of the bottom, because we already ruled out these special cases.

File: src/handlers.rs
Comment (flokli) [reply src/handlers.rs:546]: Yes, this shoudl work. Updated.

**Comment by flokli** [comment]
> The whole thing with the sentinel error-case hostent getting returned from gethostbyaddr_r is very odd to me. We don't have to reproduce libc's strangeness exactly, even if we do have to serialize that error value and send it back to the caller.

We're just forwarding `herrno`, or setting it to `0` (in the case we documented in the last push). We don't get much from trying to interpret it and carving out specific enums.

Some iteration of PR tried to do this, but turned out to be very error-prone, for little added value. We opted to stick more to how it's done by glibc here, especially considering how much userspace relies on this to be identical.

**Comment by blinsay** [comment]
> Some iteration of PR tried to do this, but turned out to be very error-prone, for little added value. We opted to stick more to how it's done by glibc here, especially considering how much userspace relies on this to be identical

What parts were more error prone? Is one of those iterations a closed PR I can go look at?

**Comment by picnoir** [comment]
> Is one of those iterations a closed PR I can go look at?

Sure. https://github.com/nix-community/nsncd/blob/15133dd5bfbbc1111afb929a7005e8f672cdce68/src/ffi.rs#L56

Don't use that, it's not a sound approach.

> What parts were more error prone?

Trying to interpret herrno ourselves and return an error hostent if `herrno != 0`. The correct thing to do here is:

1. if `hostent != null`, set herrno to 0 and return the received hostent as it is.
2. if `hostent == null`, return the error header, forward herrno.

We're not logging anything about herrno, there's no reason trying to parse it. It can only yield to more errors.

In the grand scheme of things, Nscd is a glorified (and lossy) glibc proxy. We just forward what glibc throws at us. I think that architecturally-wise, we should try to parse as few things as possible. To be franc, I even came to believe we'd be better off not even trying to parse the string bytes into a Rust string. There are valid cases where a valid glibc string could be an invalid Rust string.

**Comment by flokli** [comment]
@picnoir good call, I pushed another commit on top that now keeps these CStrings CStrings, allowing to remove some of the unnecessary back-and-forth parsing.

**Comment by blinsay** [comment]
> I pushed another commit on top that now keeps these CStrings CStrings, allowing to remove some of the unnecessary back-and-forth parsing and serializing.

Nice :D

> We're not logging anything about herrno, we're not using it for any logic, there's no reason trying to parse it. It can only yield to more errors.
>
> In the grand scheme of things, Nscd is a glorified (and lossy) glibc proxy. We just forward what glibc throws at us. I think that architecturally-wise, we should try to parse as few things as possible. Just the bare minimum to perform the Nscd error logic.

This is the right position to take imo, but it was our choice to bundle herrno in with `libc::hostent` fields. It seems like`Hostent` contains an `herror` field because that was easy. Point taken on not parsing values we don't have to, but I think that means the right thing to do is just not interpret the error code, rather than pretend there's no difference between a `libc::hostent` and and error.

Given that I think I I'd suggest we do something like this:

```rust
enum HostentError {
  HError(i32),
  Other(anyhow::Error),
}
```

or even like this:

```rust
enum Hostent {
  Ok{ name: CString, ... },
  Error(i32),
}
```

**Comment by picnoir** [comment]
done.

**Comment by blinsay** [comment]
Awesome, looks good. I'm good to merge if @leifwalsh is.

**Review by leifwalsh — APPROVED** [review]

**Comment by flokli** [comment]
:tada: thanks!

**Comment by leifwalsh** [comment]
thank you both!

**Pull request merged**
